src/lib/io/sol-memmap-storage.c

```diff
@@ -0,0 +1,366 @@
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "sol-memmap-storage.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "sol-buffer.h"
+#include "sol-log.h"
+#include "sol-str-slice.h"
+#include "sol-str-table.h"
+#include "sol-util.h"
+#include "sol-util-file.h"
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+
+static struct sol_ptr_vector memory_maps = SOL_PTR_VECTOR_INIT;
+static struct sol_ptr_vector checked_maps = SOL_PTR_VECTOR_INIT;
+
+static bool
+get_entry_metadata_on_map(const char *name, const struct sol_memmap_map *map, const struct sol_memmap_entry **entry, uint64_t *mask)
+{
+    uint32_t bit_size;
+
+    if (sol_str_table_ptr_lookup(map->entries, sol_str_slice_from_str(name), entry)) {
+        bit_size = (*entry)->bit_size;
+        /* No mask if bit_size equal or greater than 64. Such data should not be read as an int */
+        if (bit_size && (bit_size != (*entry)->size * 8) && bit_size < 64)
+            *mask = (((uint64_t)1 << bit_size) - 1) << (*entry)->bit_offset;
+        else
+            *mask = 0;
+
+        return true;
+    }
+
+    return false;
+}
+
+static bool
+get_entry_metadata(const char *name, const struct sol_memmap_map **map, const struct sol_memmap_entry **entry, uint64_t *mask)
+{
+    int i;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&memory_maps, *map, i) {
+        if (get_entry_metadata_on_map(name, *map, entry, mask))
+            return true;
+    }
+
+    entry = NULL;
+    map = NULL;
+
+    return false;
+}
+
+static int
+sol_memmap_read_raw_do(const char *path, const struct sol_memmap_entry *entry, uint64_t mask, struct sol_buffer *buffer)
+{
+    int fd, ret = 0;
+    bool add_ending_char = buffer->capacity == 0;
```

Comment:
Hmmm... @barbieri but anyway it's memmap task to add the trailing char, right? For an instant I thought that `sol_util_fill_buffer` would do it for free.